pkg/server/container_stats_list.go

```diff
@@ -17,14 +17,75 @@ limitations under the License.
 package server
 
 import (
-	"errors"
-
+	"fmt"
 	"golang.org/x/net/context"
 
+	"github.com/containerd/cgroups"
+	tasks "github.com/containerd/containerd/api/services/tasks/v1"
+	"github.com/containerd/containerd/api/types"
+	"github.com/containerd/containerd/typeurl"
 	"k8s.io/kubernetes/pkg/kubelet/apis/cri/v1alpha1/runtime"
 )
 
 // ListContainerStats returns stats of all running containers.
 func (c *criContainerdService) ListContainerStats(ctx context.Context, in *runtime.ListContainerStatsRequest) (*runtime.ListContainerStatsResponse, error) {
-	return nil, errors.New("not implemented")
+	request, err := c.buildTaskMetricsRequest(in)
+	if err != nil {
+		return nil, fmt.Errorf("failed to build metrics request: %v", err)
+	}
+	resp, err := c.taskService.Metrics(context.Background(), &request)
```

Comment:
We may want to add a filter to filter `containerKindLabel=container`. This will also include sandbox containers in the list.

I need to discuss with @yujuhong to finalize whether we should include sandbox container stats in `ListContainerStats`.

Reply:
Discussed with @yujuhong, let's exclude the sandbox container for now. We need to figure out a way to handle pod level overhead.

Reply:
ok makes sense. I wasn't sure about it.

Reply:
Circling back again. If we exclude sandbox container. The aggregated stats for a pod might be misleading right ? Or is the reason that the stats for sandbox containers is negligible and consistent for all pods so it doesnt factor into decision making ?

Reply:
@abhinandanpb This is a problem is CRI now.

We have pod level cgroup, and ideally Kubelet should make resource management decision based on resource usage got from pod level cgroup. And the container metrics are only for user who cares about how much resource their workloads (application containers) are using.

However, today Kubelet is not using pod level stats yet. It is still making decision based on container metrics. This means that the problem you mentioned does exist, kubelet is not looking at all resource usage inside the pod (sandbox container, container-shim etc.).

This is an issue needs to be addressed. We should:
1) Either change Kubelet to make resource management decision based on pod level stats, which  include everything inside the pod cgroup.
2) Or change CRI to include pod sandbox extra overhead.

1) is preferable now, and we'll probably do it next Quarter.

For now, we could only provide application container metrics based on current CRI. We'll address this in 1.9. We are going to refactor cadvisor, and pod cgroup metrics will be part of the effort.

Reply:
Got it. Thanks for the explaination.